### PR TITLE
support incremental command navigation in the terminal accessible buffer

### LINKED ISSUE #179272: [Accessibility]: Add shortcut keys to jump between executed commands in terminal buffer
CC @meganrogge

NVDA has a really great example model on this feature. I recommend experiencing this:

1. Load NVDA.

1. Press Ctrl+NVDA+Z to open NVDA's built-in Python repl.

1. Type and run multiple Python commands, such as `print("hello"); print("world");` (execute separately).
1. Press F6 to focus in the output buffer where you can use standard arrows and navigation keys to review. This is pretty much the same as VSCode a11y terminal buffer.

1. In the output buffer, press Alt+UpArrow and Alt+DownArrow to quickly jup between executed commands.

* Note: Pressing Alt+Up/DownArrows let you move through your command history. This is somewhat similar to VSCode `Ctrl+Shift+O` behavior, but this is different b/c users do not need to open additional view.

It would be so great if Alt+Up/DownArrow keys can be used in VSCode a11y terminal buffer to implement this feature.

### PR DESCRIPTION
**Pull request by meganrogge**
fix #179272

### CODE REVIEW

**Comment by meganrogge** [comment]
I need to test on Windows because on MacOS with voiceover, just positioning the cursor isn't enough for the screen reader to read the content as I navigate. 

We might also need to set the selection here. Or can await feedback...

File: src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminal.accessibility.contribution.ts [hunk 1]
```diff
@@ -53,6 +53,10 @@ class AccessibleBufferContribution extends DisposableStore implements ITerminalC
 	async createCommandQuickPick(): Promise<IQuickPick<IQuickPickItem> | undefined> {
 		return this._accessibleBufferWidget?.createQuickPick();
 	}
+
+	navigateToCommand(type: 'next' | 'previous'): void {
+		return this._accessibleBufferWidget?.navigateToCommand(type);
+	}
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminal.accessibility.contribution.ts:59]: A const enum with items assigned to numbers 1 and 2 would result in slightly less bytes in the final bundle

File: src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminal.accessibility.contribution.ts [hunk 2]
```diff
@@ -126,3 +130,48 @@ registerTerminalAction({
 		quickPick?.show();
 	}
 });
+
+registerTerminalAction({
+	id: TerminalCommandId.AccessibleBufferGoToNextCommand,
+	title: { value: localize('workbench.action.terminal.accessibleBufferGoToNextCommand', 'Accessible Buffer Go to Next Command'), original: 'Accessible Buffer Go to Next Command' },
+	f1: true,
+	category,
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminal.accessibility.contribution.ts:138]: You can omit these as they're the default

File: src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminalAccessibleBuffer.ts [hunk 3]
```diff
@@ -166,26 +166,64 @@ export class AccessibleBufferWidget extends DisposableStore {
...
 
-	async createQuickPick(): Promise<IQuickPick<IAccessibleBufferQuickPickItem> | undefined> {
-		this._cursorPosition = withNullAsUndefined(this._editorWidget.getPosition());
+	navigateToCommand(type: 'next' | 'previous'): void {
+		const currentLine = this._editorWidget.getPosition()?.lineNumber || this._getDefaultCursorPosition()?.lineNumber;
+		const commands = this._getCommandsWithEditorLine();
+		if (!commands?.length || !currentLine) {
+			return;
+		}
+		const filteredCommands = type === 'previous' ? commands.filter(c => c.lineNumber < currentLine).sort((a, b) => b.lineNumber - a.lineNumber) : commands.filter(c => c.lineNumber > currentLine).sort((a, b) => a.lineNumber - b.lineNumber);
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminalAccessibleBuffer.ts:175]: nit: You could make this faster by using reduce, it's a pretty long line currently

File: src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminalAccessibleBuffer.ts [hunk 4]
```diff
@@ -325,3 +363,6 @@ export class AccessibleBufferWidget extends DisposableStore {
 		return model!;
 	}
 }
+
+interface ITerminalCommandInEditor { command: ITerminalCommand; lineNumber: number }
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminalAccessibleBuffer.ts:367]: To match the function above? `ICommandWithEditorLine`? Can also omit terminal as you're not exporting

**Review by Tyriar — DISMISSED** [review]

File: src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminalAccessibleBuffer.ts (hunk 3 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminalAccessibleBuffer.ts]: I always find reduce functions confusing to read and it won't be that many imo to make it worth that

File: src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminal.accessibility.contribution.ts [hunk 5]
```diff
@@ -126,3 +133,46 @@ registerTerminalAction({
...
+		await AccessibleBufferContribution.get(instance)?.navigateToCommand(NavigationType.Next);
+	}
+});
+
+
+registerTerminalAction({
+	id: TerminalCommandId.AccessibleBufferGoToPreviousCommand,
+	title: { value: localize('workbench.action.terminal.accessibleBufferGoToPreviousCommand', 'Accessible Buffer Go to Previous Command'), original: 'Accessible Buffer Go to Previous Command' },
+	f1: true,
+	category,
```
Comment (meganrogge) [inline src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminal.accessibility.contribution.ts:162]:
```suggestion
```

**Comment by meganrogge** [comment]
not sure why the tests are failing, they look unrelated

**Comment by meganrogge** [comment]
started failing here, so reverting that

https://github.com/microsoft/vscode/pull/179415/commits/1945e48fad744ce87600c007ad1a44cc12796d21

**Review by Tyriar — DISMISSED** [review]

**Comment by meganrogge** [comment]
there was a cyclic dependency

**Review by eleanorjboyd — APPROVED** [review]

**Pull request merged**